### ISSUE TITLE
Class::Struct autogenerated 'new' method causes infinite loop

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#3323]
Migrated from [rt.perl.org#5565](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=5565) (status was 'resolved')

Searchable as RT5565$

**Comment by p5pRT** [comment]
### From train1@ms.com



Consider the following sample program&#8203;:

&nbsp;    package Person;
&nbsp;    use Class&#8203;::Struct;
 
&nbsp;    struct Person => 
&nbsp;      \{
&nbsp;       name => '$'\,
&nbsp;       spouse => 'Person'\,
&nbsp;      \};

&nbsp;    package main;
&nbsp;    
&nbsp;    my $z = Person\->new;

The 'new' method generated here is&#8203;:

&nbsp;        sub new \{
&nbsp;          my\($r\) = \{\};
&nbsp;          $r\->\{'name'\} = undef;
&nbsp;          $r\->\{'spouse'\} = 'Person'\->new\(\);
&nbsp;          bless $r;
&nbsp;        \}

so that the Person&#8203;::new function calls itself in an infinite recursion\.

<details><summary>Perl Info</summary>

```


Site configuration information for perl 5.00404:

Configured by wpm at Wed Nov 26 16:55:08 EST 1997.

Summary of my perl5 (5.0 patchlevel 4 subversion 4) configuration:
  Platform:
    osname=solaris, osvers=2.5.1, archname=sun4-solaris
    uname='sunos zappa 5.5.1 generic_103640-12-1 sun4u sparc sunw,ultra-1 '
    hint=recommended, useposix=true, d_sigaction=define
    bincompat3=y useperlio=undef d_sfio=undef
  Compiler:
    cc='/ms/dist/fsf/PROJ/gcc/2.7.2.0/bin/gcc', optimize='-O', gccversion=2.7.2
    cppflags='-DDEBUGGING_MSTATS -DPERL_EMERGENCY_SBRK -DPACK_MALLOC -DTWO_POT_OPTIMIZE -I/ms/dist/aurora/include -I/ms/dist/fsf/include'
    ccflags ='-DDEBUGGING_MSTATS -DPERL_EMERGENCY_SBRK -DPACK_MALLOC -DTWO_POT_OPTIMIZE -I/ms/dist/aurora/include -I/ms/dist/fsf/include'
    stdchar='unsigned char', d_stdstdio=define, usevfork=false
    voidflags=15, castflags=0, d_casti32=define, d_castneg=define
    intsize=4, alignbytes=8, usemymalloc=y, prototype=define
  Linker and Libraries:
    ld='/ms/dist/fsf/PROJ/gcc/2.7.2.0/bin/gcc', ldflags =' -L/ms/dist/aurora/lib -L/ms/dist/fsf/lib'
    libpth=/ms/dist/aurora/lib /ms/dist/fsf/lib /lib /usr/lib /usr/ccs/lib /usr/local/lib
    libs=-lsocket -lnsl -lgdbm -ldb -ldl -lm -lc -lcrypt
    libc=/lib/libc.so, so=so
    useshrplib=false, libperl=libperl.a
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags=' '
    cccdlflags='-fPIC', lddlflags='-G -L/ms/dist/aurora/lib -L/ms/dist/fsf/lib'

Locally applied patches:
	


@INC for perl 5.00404:
	/ms/dist/perl5/PROJ/core/5.00404-n32/exec/lib/perl5
	/ms/dist/perl5/PROJ/core/5.00404-n32/common/lib/perl5
	/ms/dist/perl5/VERS/5.004/lib/perl5
	/ms/dist/perl5/lib/perl5
	/ms/dist/aurora/lib/perl5
	.


Environment for perl 5.00404:
    HOME=/ms/user/t/train1
    LANG=C
    LD_LIBRARY_PATH=/usr/dt/lib:/opt/softbench/lib:/usr/openwin/lib:/ms/dist/aurora/lib
    LOGDIR (unset)
    PATH=/ms/user/t/train1/bin:/ms/dist/aurora/hackbin:/ms/dist/aurora/bin:/bin:/usr/ccs/bin:/usr/ucb:/ms/dist/fsf/bin:/ms/dist/afs/bin:/ms/dist/aurora/sbin:/usr/sbin:/ms/dist/afs/sbin:.:/usr/dt/bin:/usr/openwin/bin
    PERL_BADLANG (unset)
    SHELL=/bin/ksh

```
</details>

**Comment by p5pRT** [comment]
### From @rspier

Appears to have been fixed in Class&#8203;::Struct v0\.62\.  \(Not repeatable in
perl5\.8\.0\)

&nbsp;              \* "recursive" data structures didnt work well&#8203;:
&nbsp;                        package Person;
&nbsp;                        struct \{ mother => Person\, father => Person\};

**Comment by p5pRT** [comment]
@rspier - Status changed from 'open' to 'resolved'
